arangod/RocksDBEngine/Methods/RocksDBTrxMethods.cpp

```diff
@@ -202,6 +244,138 @@ Result RocksDBTrxMethods::checkIntermediateCommit(uint64_t newSize, bool& hasPer
+    // don't bother with query tracking in non globally managed trx
+    return;
+  }
+
+  if (isModificationQuery) {
+    TRI_ASSERT(_readWriteBatch != nullptr);
+    TRI_ASSERT(_hasActiveModificationQuery.load());
+    TRI_ASSERT(_numActiveReadOnlyQueries.load() == 0);
+    _hasActiveModificationQuery.store(false, std::memory_order_relaxed);
+    releaseReadWriteBatch();
+    TRI_ASSERT(_readWriteBatch == nullptr && !_ownsReadWriteBatch);
+    _readWriteBatch = _rocksTransaction->GetWriteBatch();
+  } else {
+    TRI_ASSERT(_ownsReadWriteBatch == false);
+    TRI_ASSERT(_hasActiveModificationQuery.load() == false);
+    TRI_ASSERT(_numActiveReadOnlyQueries.load() > 0);
+    _numActiveReadOnlyQueries.fetch_sub(1, std::memory_order_relaxed);
+  }
+}
+
+void RocksDBTrxMethods::initializeReadWriteBatch() {
+  TRI_ASSERT(_ownsReadWriteBatch == false);
+  _readWriteBatch = new rocksdb::WriteBatchWithIndex(rocksdb::BytewiseComparator(), 0, true, 0);
+  _ownsReadWriteBatch = true;
+
+  struct WBCloner final : public rocksdb::WriteBatch::Handler {
+    explicit WBCloner(rocksdb::WriteBatchWithIndex& target)
+        : wbwi(target) {}
+
+    rocksdb::Status PutCF(uint32_t column_family_id, const rocksdb::Slice& key,
+                          const rocksdb::Slice& value) override {
+      return wbwi.Put(familyId(column_family_id), key, value);
+    }
+
+    rocksdb::Status DeleteCF(uint32_t column_family_id, const rocksdb::Slice& key) override {
+      return wbwi.Delete(familyId(column_family_id), key);
+    }
+
+    rocksdb::Status SingleDeleteCF(uint32_t column_family_id, const rocksdb::Slice& key) override {
+      return wbwi.SingleDelete(familyId(column_family_id), key);
+    }
+
+    rocksdb::Status DeleteRangeCF(uint32_t column_family_id, const rocksdb::Slice& begin_key,
+                                  const rocksdb::Slice& end_key) override {
+      return wbwi.DeleteRange(familyId(column_family_id), begin_key, end_key);
+    }
+
+    rocksdb::Status MergeCF(uint32_t column_family_id, const rocksdb::Slice& key,
+                            const rocksdb::Slice& value) override {
+      // should never be used by our code
+      TRI_ASSERT(false);
+      return wbwi.Merge(familyId(column_family_id), key, value);
+    }
+
+    void LogData(const rocksdb::Slice& blob) override {
+      wbwi.PutLogData(blob);
+    }
+
+    rocksdb::ColumnFamilyHandle* familyId(uint32_t id) {
+      for (auto const& it : RocksDBColumnFamilyManager::allHandles()) {
+        if (it->GetID() == id) {
+          return it;
+        }
+      }
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "unknown column family id");
+    }
+
+    rocksdb::WriteBatchWithIndex& wbwi;
+  } cloner(*_readWriteBatch);
+
+  // get the transaction's current WriteBatch
+  TRI_ASSERT(_rocksTransaction != nullptr);
+  rocksdb::WriteBatch* wb = _rocksTransaction->GetWriteBatch()->GetWriteBatch();
+  rocksdb::Status s = wb->Iterate(&cloner);
```

Comment:
Yes, but this would probably be rather hacky. We have to access the WriteBatch's underlying data buffer (` rep_` , which is protected), but the only getter returns a const&. So we either have to use a const_cast or use some hack to access the protected member. Not sure if we want to go down that road... I would probably leave it like this and run some performance tests later to compare the clean and the hacky solutions.